File: usages/collector/collector.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"time"
)

type CIActionFile struct {
	Path string `json:"path"`
}

type Usage struct {
	CIActionFile string `json:"ci_action_file"`
	RepoName     string `json:"repo_name"`
	RepoFile     string `json:"repo_file"`
}

type CollectedData struct {
	AllCIActionFiles []CIActionFile `json:"all_ci_action_files"`
	Usages           []Usage        `json:"usages"`
}

type GitHubRepo struct {
	Name     string `json:"name"`
	Archived bool   `json:"archived"`
}

type GitHubContent struct {
	Name        string `json:"name"`
	Path        string `json:"path"`
	Type        string `json:"type"`
	DownloadURL string `json:"download_url"`
}

var (
	usesActionRegex   = regexp.MustCompile(`uses:\s+untillpro/ci-action@(\S+)`)
	usesWorkflowRegex = regexp.MustCompile(`uses:\s+untillpro/ci-action/(\.github/workflows/[^@]+)@(\S+)`)
	curlScriptRegex   = regexp.MustCompile(`https://raw\.githubusercontent\.com/untillpro/ci-action/(\S+?)/scripts/([^\s|]+)`)
	curlAnyRegex      = regexp.MustCompile(`https://raw\.githubusercontent\.com/untillpro/ci-action/(\S+?)/([^\s|]+)`)
	httpClient        = &http.Client{Timeout: 30 * time.Second}
)

const (
	githubAPIBase = "https://api.github.com"
	orgName       = "untillpro"
)

func main() {
	workDir, err := os.Getwd()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error getting working directory: %v\n", err)
		os.Exit(1)
	}

	rootDir := filepath.Join(workDir, "..")
	ciActionPath := filepath.Join(rootDir, "ci-action")
	var allCIActionFiles []CIActionFile

	if _, err := os.Stat(ciActionPath); os.IsNotExist(err) {
		fmt.Println("Local ci-action folder not found, fetching from GitHub...")
		allCIActionFiles, err = getAllCIActionFilesFromGitHub()
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error fetching ci-action files from GitHub: %v\n", err)
			os.Exit(1)
		}
	} else {
		allCIActionFiles = getAllCIActionFiles(ciActionPath)
	}

	outdatedReposFile := filepath.Join(rootDir, "outdated-repos.txt")
	outdatedRepos, err := loadOutdatedRepos(outdatedReposFile)
	if err != nil {
		outdatedRepos = make(map[string]bool)
	} else if len(outdatedRepos) > 0 {
		fmt.Printf("Loaded %d outdated repositories to skip\n", len(outdatedRepos))
	}

	fmt.Println("Fetching non-archived repositories from GitHub...")
	repos, err := fetchNonArchivedRepos()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error fetching repositories: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("Found %d non-archived repositories\n", len(repos))

	var usages []Usage
	reposWithUsages := make(map[string]bool)
	skippedCount := 0

	for idx := range repos {
		repoName := repos[idx].Name
		fullRepoName := orgName + "/" + repoName

		if outdatedRepos[fullRepoName] || outdatedRepos[repoName] {
			skippedCount++
			continue
		}

		fmt.Printf("Scanning %s (%d/%d)...\n", repoName, idx+1, len(repos))

		repoUsages, err := scanRepositoryFromGitHub(repoName)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Warning: Error scanning %s: %v\n", repoName, err)
			continue
		}

		if len(repoUsages) > 0 {
			usages = append(usages, repoUsages...)
			reposWithUsages[repoName] = true
		}
	}

	if skippedCount > 0 {
		fmt.Printf("Skipped %d outdated repositories\n", skippedCount)
	}

	data := CollectedData{
		AllCIActionFiles: allCIActionFiles,
		Usages:           usages,
	}

	jsonData, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error marshaling JSON: %v\n", err)
		os.Exit(1)
	}

	outputFile := filepath.Join(rootDir, "ci-action-data.json")
	if err := os.WriteFile(outputFile, jsonData, 0644); err != nil {
		fmt.Fprintf(os.Stderr, "Error writing JSON file: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("\nCollected %d ci-action files and %d usages from %d repositories\n",
		len(allCIActionFiles), len(usages), len(reposWithUsages))
	fmt.Println("Data written to ci-action-data.json")
}

func loadOutdatedRepos(filename string) (map[string]bool, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	outdated := make(map[string]bool)
	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line != "" && !strings.HasPrefix(line, "#") {
			outdated[line] = true
		}
	}

	if err := scanner.Err(); err != nil {
		return nil, err
	}

	return outdated, nil
}

func fetchNonArchivedRepos() ([]GitHubRepo, error) {
	var allRepos []GitHubRepo
	page := 1
	githubToken := os.Getenv("GITHUB_TOKEN")

	for {
		url := fmt.Sprintf("%s/orgs/%s/repos?per_page=100&page=%d&type=all", githubAPIBase, orgName, page)
		req, err := http.NewRequest("GET", url, nil)
		if err != nil {
			return nil, err
		}

		req.Header.Set("Accept", "application/vnd.github.v3+json")
		if githubToken != "" {
			req.Header.Set("Authorization", "Bearer "+githubToken)
		}

		resp, err := httpClient.Do(req)
		if err != nil {
			return nil, err
		}

		if resp.StatusCode != http.StatusOK {
			resp.Body.Close()
			return nil, fmt.Errorf("GitHub API returned status %d", resp.StatusCode)
		}

		var repos []GitHubRepo
		if err := json.NewDecoder(resp.Body).Decode(&repos); err != nil {
			resp.Body.Close()
			return nil, err
		}
		resp.Body.Close()

		if len(repos) == 0 {
			break
		}

		for idx := range repos {
			if !repos[idx].Archived {
				allRepos = append(allRepos, repos[idx])
			}
		}

		page++
		time.Sleep(100 * time.Millisecond)
	}

	return allRepos, nil
}

func scanRepositoryFromGitHub(repoName string) ([]Usage, error) {
	contents, err := fetchGitHubDirectoryContents(repoName, ".github")
	if err != nil {
		return nil, nil
	}

	var usages []Usage

	for idx := range contents {
		if contents[idx].Type == "file" {
			fileUsages, err := scanGitHubFile(repoName, contents[idx].Path, contents[idx].DownloadURL)
			if err != nil {
				continue
			}
			usages = append(usages, fileUsages...)
		} else if contents[idx].Type == "dir" {
			subUsages, err := scanGitHubDirectory(repoName, contents[idx].Path)
			if err != nil {
				continue
			}
			usages = append(usages, subUsages...)
		}
	}

	return usages, nil
}

func scanGitHubDirectory(repoName, dirPath string) ([]Usage, error) {
	contents, err := fetchGitHubDirectoryContents(repoName, dirPath)
	if err != nil {
		return nil, err
	}

	var usages []Usage

	for idx := range contents {
		if contents[idx].Type == "file" {
			fileUsages, err := scanGitHubFile(repoName, contents[idx].Path, contents[idx].DownloadURL)
			if err != nil {
				continue
			}
			usages = append(usages, fileUsages...)
		} else if contents[idx].Type == "dir" {
			subUsages, err := scanGitHubDirectory(repoName, contents[idx].Path)
			if err != nil {
				continue
			}
			usages = append(usages, subUsages...)
		}
	}

	return usages, nil
}

func fetchGitHubDirectoryContents(repoName, dirPath string) ([]GitHubContent, error) {
	url := fmt.Sprintf("%s/repos/%s/%s/contents/%s", githubAPIBase, orgName, repoName, dirPath)
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, err
	}

	req.Header.Set("Accept", "application/vnd.github.v3+json")
	githubToken := os.Getenv("GITHUB_TOKEN")
	if githubToken != "" {
		req.Header.Set("Authorization", "Bearer "+githubToken)
	}

	resp, err := httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusNotFound {
		return nil, fmt.Errorf("directory not found")
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("GitHub API returned status %d", resp.StatusCode)
	}

	var contents []GitHubContent
	if err := json.NewDecoder(resp.Body).Decode(&contents); err != nil {
		return nil, err
	}

	time.Sleep(100 * time.Millisecond)

	return contents, nil
}

func scanGitHubFile(repoName, filePath, downloadURL string) ([]Usage, error) {
	resp, err := httpClient.Get(downloadURL)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("failed to download file")
	}

	var usages []Usage
	scanner := bufio.NewScanner(resp.Body)

	for scanner.Scan() {
		line := scanner.Text()
		fileUsages := extractUsagesFromLine(line, repoName, filePath)
		usages = append(usages, fileUsages...)
	}

	time.Sleep(50 * time.Millisecond)

	return usages, nil
}

func getAllCIActionFilesFromGitHub() ([]CIActionFile, error) {
	var files []CIActionFile
	githubToken := os.Getenv("GITHUB_TOKEN")

	if _, err := os.Stat(filepath.Join("ci-action", "action.yml")); err == nil {
		files = append(files, CIActionFile{Path: "action.yml"})
	}

	workflowsContents, err := fetchGitHubDirectoryContents("ci-action", ".github/workflows")
	if err == nil {
		for idx := range workflowsContents {
			if workflowsContents[idx].Type == "file" && strings.HasSuffix(workflowsContents[idx].Name, ".yml") {
				files = append(files, CIActionFile{Path: ".github/workflows/" + workflowsContents[idx].Name})
			}
		}
	}

	scriptsContents, err := fetchGitHubDirectoryContents("ci-action", "scripts")
	if err == nil {
		for idx := range scriptsContents {
			if scriptsContents[idx].Type == "file" && strings.HasSuffix(scriptsContents[idx].Name, ".sh") {
				files = append(files, CIActionFile{Path: "scripts/" + scriptsContents[idx].Name})
			}
		}
	}

	actionURL := fmt.Sprintf("%s/repos/%s/ci-action/contents/action.yml", githubAPIBase, orgName)
	req, err := http.NewRequest("GET", actionURL, nil)
	if err == nil {
		req.Header.Set("Accept", "application/vnd.github.v3+json")
		if githubToken != "" {
			req.Header.Set("Authorization", "Bearer "+githubToken)
		}
		resp, err := httpClient.Do(req)
		if err == nil && resp.StatusCode == http.StatusOK {
			resp.Body.Close()
			files = append(files, CIActionFile{Path: "action.yml"})
		} else if resp != nil {
			resp.Body.Close()
		}
	}

	sort.Slice(files, func(i, j int) bool {
		return files[i].Path < files[j].Path
	})

	uniqueFiles := make([]CIActionFile, 0)
	seen := make(map[string]bool)
	for idx := range files {
		if !seen[files[idx].Path] {
			uniqueFiles = append(uniqueFiles, files[idx])
			seen[files[idx].Path] = true
		}
	}

	return uniqueFiles, nil
}

func getAllCIActionFiles(ciActionPath string) []CIActionFile {
	var files []CIActionFile

	if _, err := os.Stat(filepath.Join(ciActionPath, "action.yml")); err == nil {
		files = append(files, CIActionFile{Path: "action.yml"})
	}

	workflowsPath := filepath.Join(ciActionPath, ".github", "workflows")
	if entries, err := os.ReadDir(workflowsPath); err == nil {
		for _, entry := range entries {
			if !entry.IsDir() && strings.HasSuffix(entry.Name(), ".yml") {
				files = append(files, CIActionFile{Path: ".github/workflows/" + entry.Name()})
			}
		}
	}

	scriptsPath := filepath.Join(ciActionPath, "scripts")
	if entries, err := os.ReadDir(scriptsPath); err == nil {
		for _, entry := range entries {
			if !entry.IsDir() && strings.HasSuffix(entry.Name(), ".sh") {
				files = append(files, CIActionFile{Path: "scripts/" + entry.Name()})
			}
		}
	}

	sort.Slice(files, func(i, j int) bool {
		return files[i].Path < files[j].Path
	})

	return files
}

func extractUsagesFromLine(line, repoName, relPath string) []Usage {
	var usages []Usage

	if matches := usesWorkflowRegex.FindStringSubmatch(line); matches != nil {
		workflowPath := matches[1]
		usages = append(usages, Usage{
			CIActionFile: workflowPath,
			RepoName:     repoName,
			RepoFile:     relPath,
		})
	}

	if matches := usesActionRegex.FindStringSubmatch(line); matches != nil {
		if !usesWorkflowRegex.MatchString(line) {
			usages = append(usages, Usage{
				CIActionFile: "action.yml",
				RepoName:     repoName,
				RepoFile:     relPath,
			})
		}
	}

	if matches := curlScriptRegex.FindStringSubmatch(line); matches != nil {
		scriptName := matches[2]
		usages = append(usages, Usage{
			CIActionFile: "scripts/" + scriptName,
			RepoName:     repoName,
			RepoFile:     relPath,
		})
	} else if matches := curlAnyRegex.FindStringSubmatch(line); matches != nil {
		if !strings.Contains(line, "/scripts/") {
			filePath := matches[2]
			usages = append(usages, Usage{
				CIActionFile: filePath,
				RepoName:     repoName,
				RepoFile:     relPath,
			})
		}
	}

	return usages
}

File: docs/scripts/collector.py
#!/usr/bin/env python3
"""
CI-Action Usage Collector

Scans all non-archived repositories from GitHub and collects raw usage data.
"""

import csv
import json
import os
import re
import sys
import time
from pathlib import Path
from typing import Dict, List, Optional
from urllib.request import Request, urlopen
from urllib.error import HTTPError, URLError


# Regular expressions for detecting ci-action usage
USES_ACTION_REGEX = re.compile(r'uses:\s+untillpro/ci-action@(\S+)')
USES_WORKFLOW_REGEX = re.compile(r'uses:\s+untillpro/ci-action/(\.github/workflows/[^@]+)@(\S+)')
CURL_SCRIPT_REGEX = re.compile(r'https://raw\.githubusercontent\.com/untillpro/ci-action/(\S+?)/scripts/([^\s|]+)')
CURL_ANY_REGEX = re.compile(r'https://raw\.githubusercontent\.com/untillpro/ci-action/(\S+?)/([^\s|]+)')
# Match $GITHUB_ACTION_PATH/../scripts/script.sh pattern (with or without quotes)
ACTION_PATH_SCRIPT_REGEX = re.compile(r'\$GITHUB_ACTION_PATH/\.\./scripts/([^\s"\']+)')

# Constants
GITHUB_API_BASE = "https://api.github.com"
ORG_NAME = "untillpro"
REQUEST_TIMEOUT = 30

# Repositories to skip
SKIP_REPOS = {"untillpro/voedger"}

# Additional repositories to scan (outside untillpro org)
ADDITIONAL_REPOS = [
    {"name": "voedger", "owner": "voedger", "default_branch": "main"}
]


class GitHubClient:
    """GitHub API client with authentication and rate limiting."""

    def __init__(self, token: Optional[str] = None):
        self.token = token or os.getenv("GITHUB_TOKEN")

    def get(self, url: str) -> Dict:
        """Make GET request with error handling."""
        req = Request(url)
        req.add_header("Accept", "application/vnd.github.v3+json")
        if self.token:
            req.add_header("Authorization", f"Bearer {self.token}")

        try:
            with urlopen(req, timeout=REQUEST_TIMEOUT) as response:
                return {
                    'status_code': response.status,
                    'data': json.loads(response.read().decode('utf-8'))
                }
        except HTTPError as e:
            return {
                'status_code': e.code,
                'data': None
            }
        except URLError:
            return {
                'status_code': 0,
                'data': None
            }


def load_outdated_repos(filename: str) -> set:
    """Load list of outdated repositories to skip."""
    outdated = set()
    try:
        with open(filename, 'r') as f:
            for line in f:
                line = line.strip()
                if line and not line.startswith('#'):
                    outdated.add(line)
    except FileNotFoundError:
        pass
    return outdated


def fetch_non_archived_repos(client: GitHubClient) -> List[Dict]:
    """Fetch all non-archived repositories from the organization."""
    all_repos = []
    page = 1

    while True:
        url = f"{GITHUB_API_BASE}/orgs/{ORG_NAME}/repos?per_page=100&page={page}&type=all"
        response = client.get(url)

        if response['status_code'] != 200:
            raise Exception(f"GitHub API returned status {response['status_code']}")

        repos = response['data']
        if not repos:
            break

        for repo in repos:
            if not repo.get('archived', False):
                all_repos.append({
                    'name': repo['name'],
                    'owner': repo['owner']['login'],
                    'default_branch': repo.get('default_branch', 'main')
                })

        page += 1
        time.sleep(0.1)  # Rate limiting

    return all_repos


def fetch_github_directory_contents(client: GitHubClient, owner: str, repo_name: str, dir_path: str) -> Optional[List[Dict]]:
    """Fetch contents of a directory from GitHub."""
    url = f"{GITHUB_API_BASE}/repos/{owner}/{repo_name}/contents/{dir_path}"
    response = client.get(url)

    if response['status_code'] == 404:
        return None

    if response['status_code'] != 200:
        return None

    time.sleep(0.1)  # Rate limiting
    return response['data']


def extract_usages_from_line(line: str, repo_info: Dict, file_path: str, line_number: int) -> List[Dict]:
    """Extract ci-action usages from a single line of text."""
    usages = []

    # Build the source URL once
    source_url = f"https://github.com/{repo_info['owner']}/{repo_info['name']}/blob/{repo_info['default_branch'] or 'main'}/{file_path}#L{line_number}"

    # Check for workflow usage (most specific pattern first)
    workflow_match = USES_WORKFLOW_REGEX.search(line)
    if workflow_match:
        usages.append({
            'ci_action_file': workflow_match.group(1),
            'source_url': source_url
        })
    # Check for action usage (only if not a workflow usage)
    elif USES_ACTION_REGEX.search(line):
        usages.append({
            'ci_action_file': 'action.yml',
            'source_url': source_url
        })

    # Check for script curl usage
    script_match = CURL_SCRIPT_REGEX.search(line)
    if script_match:
        usages.append({
            'ci_action_file': f'scripts/{script_match.group(2)}',
            'source_url': source_url
        })
    # Check for any other curl usage (only if no script match)
    elif '/scripts/' not in line:
        any_match = CURL_ANY_REGEX.search(line)
        if any_match:
            usages.append({
                'ci_action_file': any_match.group(2),
                'source_url': source_url
            })

    # Check for $GITHUB_ACTION_PATH/../scripts/ usage
    action_path_match = ACTION_PATH_SCRIPT_REGEX.search(line)
    if action_path_match:
        usages.append({
            'ci_action_file': f'scripts/{action_path_match.group(1)}',
            'source_url': source_url
        })

    return usages


def scan_github_file(client: GitHubClient, repo_info: Dict, file_path: str, download_url: str) -> List[Dict]:
    """Scan a GitHub file for ci-action usages."""
    try:
        req = Request(download_url)
        with urlopen(req, timeout=REQUEST_TIMEOUT) as response:
            if response.status != 200:
                return []

            content = response.read().decode('utf-8')
            usages = []
            for line_number, line in enumerate(content.splitlines(), start=1):
                usages.extend(extract_usages_from_line(line, repo_info, file_path, line_number))

            time.sleep(0.05)  # Rate limiting
            return usages
    except Exception:
        return []


def scan_github_directory(client: GitHubClient, repo_info: Dict, dir_path: str) -> List[Dict]:
    """Recursively scan a GitHub directory for ci-action usages."""
    contents = fetch_github_directory_contents(client, repo_info['owner'], repo_info['name'], dir_path)
    if not contents:
        return []

    usages = []
    for item in contents:
        if item['type'] == 'file':
            # Skip markdown files
            if item['name'].endswith('.md'):
                continue
            file_usages = scan_github_file(client, repo_info, item['path'], item['download_url'])
            usages.extend(file_usages)
        elif item['type'] == 'dir':
            sub_usages = scan_github_directory(client, repo_info, item['path'])
            usages.extend(sub_usages)

    return usages


def scan_repository_from_github(client: GitHubClient, repo_info: Dict) -> List[Dict]:
    """Scan a repository for ci-action usages."""
    return scan_github_directory(client, repo_info, '.github')


def get_all_ci_action_files_from_github(client: GitHubClient) -> List[Dict]:
    """Get all ci-action files from the ci-action repository."""
    files = {}  # Use dict to automatically handle duplicates

    # Check for checkout-and-setup-go/action.yml
    url = f"{GITHUB_API_BASE}/repos/{ORG_NAME}/ci-action/contents/checkout-and-setup-go/action.yml"
    response = client.get(url)
    if response['status_code'] == 200:
        files['checkout-and-setup-go/action.yml'] = {'path': 'checkout-and-setup-go/action.yml'}

    # Get workflow files
    workflows = fetch_github_directory_contents(client, ORG_NAME, 'ci-action', '.github/workflows')
    if workflows:
        for item in workflows:
            if item['type'] == 'file' and item['name'].endswith('.yml'):
                path = f".github/workflows/{item['name']}"
                files[path] = {'path': path}

    # Get script files
    scripts = fetch_github_directory_contents(client, ORG_NAME, 'ci-action', 'scripts')
    if scripts:
        for item in scripts:
            if item['type'] == 'file' and item['name'].endswith('.sh'):
                path = f"scripts/{item['name']}"
                files[path] = {'path': path}

    # Return sorted list
    return sorted(files.values(), key=lambda x: x['path'])


def get_all_ci_action_files_local(ci_action_path: Path) -> List[Dict]:
    """Get all ci-action files from local ci-action repository."""
    files = []

    # Check for checkout-and-setup-go/action.yml
    if (ci_action_path / 'checkout-and-setup-go' / 'action.yml').exists():
        files.append({'path': 'checkout-and-setup-go/action.yml'})

    # Get workflow files
    workflows_path = ci_action_path / '.github' / 'workflows'
    if workflows_path.exists():
        for file in workflows_path.glob('*.yml'):
            files.append({'path': f'.github/workflows/{file.name}'})

    # Get script files
    scripts_path = ci_action_path / 'scripts'
    if scripts_path.exists():
        for file in scripts_path.glob('*.sh'):
            files.append({'path': f'scripts/{file.name}'})

    return sorted(files, key=lambda x: x['path'])


def scan_ci_action_internal_usages_local(ci_action_path: Path) -> List[Dict]:
    """Scan ci-action's own files for internal script usages."""
    usages = []
    repo_info = {'owner': ORG_NAME, 'name': 'ci-action', 'default_branch': 'main'}

    # Scan checkout-and-setup-go/action.yml
    action_file = ci_action_path / 'checkout-and-setup-go' / 'action.yml'
    if action_file.exists():
        content = action_file.read_text(encoding='utf-8')
        for line_number, line in enumerate(content.splitlines(), start=1):
            line_usages = extract_usages_from_line(line, repo_info, 'checkout-and-setup-go/action.yml', line_number)
            usages.extend(line_usages)

    return usages


def scan_ci_action_internal_usages_github(client: GitHubClient) -> List[Dict]:
    """Scan ci-action's own files for internal script usages from GitHub."""
    usages = []
    repo_info = {'owner': ORG_NAME, 'name': 'ci-action', 'default_branch': 'main'}

    # Scan checkout-and-setup-go/action.yml
    url = f"{GITHUB_API_BASE}/repos/{ORG_NAME}/ci-action/contents/checkout-and-setup-go/action.yml"
    response = client.get(url)
    if response['status_code'] == 200:
        import base64
        content = base64.b64decode(response['data']['content']).decode('utf-8')
        for line_number, line in enumerate(content.splitlines(), start=1):
            line_usages = extract_usages_from_line(line, repo_info, 'checkout-and-setup-go/action.yml', line_number)
            usages.extend(line_usages)

    return usages


def main():
    """Main entry point."""
    work_dir = Path.cwd()
    ci_action_path = work_dir.parent

    # Initialize GitHub client
    client = GitHubClient()

    # Get all ci-action files and scan for internal usages
    if (ci_action_path / 'checkout-and-setup-go' / 'action.yml').exists():
        print("Using local ci-action folder...")
        all_ci_action_files = get_all_ci_action_files_local(ci_action_path)
        internal_usages = scan_ci_action_internal_usages_local(ci_action_path)
    else:
        print("Local ci-action folder not found, fetching from GitHub...")
        all_ci_action_files = get_all_ci_action_files_from_github(client)
        internal_usages = scan_ci_action_internal_usages_github(client)

    # Load outdated repos
    outdated_repos_file = work_dir / 'outdated-repos.txt'
    outdated_repos = load_outdated_repos(str(outdated_repos_file))
    if outdated_repos:
        print(f"Loaded {len(outdated_repos)} outdated repositories to skip")

    # Fetch repositories
    print("Fetching non-archived repositories from GitHub...")
    repos = fetch_non_archived_repos(client)
    print(f"Found {len(repos)} non-archived repositories")

    # Scan repositories
    usages = list(internal_usages)  # Start with internal ci-action usages
    repos_with_usages = set()
    skipped_count = 0

    for idx, repo in enumerate(repos, 1):
        repo_name = repo['name']
        full_repo_name = f"{repo['owner']}/{repo_name}"

        if repo_name in outdated_repos or full_repo_name in outdated_repos:
            skipped_count += 1
            continue

        if full_repo_name in SKIP_REPOS:
            skipped_count += 1
            continue

        print(f"Scanning {full_repo_name} ({idx}/{len(repos)})...")

        try:
            repo_usages = scan_repository_from_github(client, repo)
            if repo_usages:
                usages.extend(repo_usages)
                repos_with_usages.add(full_repo_name)
        except Exception as e:
            print(f"Warning: Error scanning {full_repo_name}: {e}", file=sys.stderr)

    if skipped_count > 0:
        print(f"Skipped {skipped_count} outdated repositories")

    # Scan additional repositories
    for repo in ADDITIONAL_REPOS:
        full_repo_name = f"{repo['owner']}/{repo['name']}"
        print(f"Scanning additional repo {full_repo_name}...")
        try:
            repo_usages = scan_repository_from_github(client, repo)
            if repo_usages:
                usages.extend(repo_usages)
                repos_with_usages.add(full_repo_name)
        except Exception as e:
            print(f"Warning: Error scanning {full_repo_name}: {e}", file=sys.stderr)

    # Find unused ci-action files and add them to usages with source_url=null
    used_files = set()
    for usage in usages:
        used_files.add(usage['ci_action_file'])

    for ci_file in all_ci_action_files:
        if ci_file['path'] not in used_files:
            usages.append({
                'ci_action_file': ci_file['path'],
                'source_url': None
            })

    # Write CSV output
    output_file = work_dir / 'ci-action-data.csv'
    with open(output_file, 'w', newline='', encoding='utf-8') as f:
        writer = csv.DictWriter(f, fieldnames=['ci_action_file', 'source_url'])
        writer.writeheader()
        writer.writerows(usages)

    print(f"\nCollected {len(all_ci_action_files)} ci-action files and {len(usages)} usages from {len(repos_with_usages)} repositories")
    print("Data written to ci-action-data.csv")


if __name__ == '__main__':
    main()

